language/ankush-playground/src/Vote.move

```diff
@@ -0,0 +1,260 @@
+module DiemFramework::Vote {
+
+    use Std::Option::{Self, Option};
+    use Std::Signer;
+    use Std::Vector;
+    // use DiemFramework::DiemTimestamp;
+    // use DiemFramework::Roles;
+
+    const MAX_BALLOTS_PER_PROPOSAL_PER_ADDRESS: u64 = 256;
+
+    const ETOO_MANY_BALLOTS: u64 = 1;
+    const EBALLOT_NOT_FOUND: u64 = 2;
+    const EPROPOSAL_MISMATCH: u64 = 3;
+    const EPROPOSAL_TYPE_MISMATCH: u64 = 4;
+    const EAPPROVAL_POLICY_MISMATCH: u64 = 5;
+    const EVOTER_NOT_ALLOWED: u64 = 6;
+    const EVOTING_NOT_STARTED: u64 = 7;
+    const EVOTING_EXPIRED: u64 = 8;
+    const EALREADY_VOTED: u64 = 9;
+    const EBALLOT_ALREADY_COMPLETE: u64 = 10;
+    const EINVALID_APPROVAL_POLICY: u64 = 11;
+    const EINVALID_BALLOT_TIMESTAMPS: u64 = 12;
+    const EAPPROVAL_POLICES_EXIST: u64 = 13;
+    const ETODO: u64 = 14; // TODO: Add appropriate error
+
+
+    // ApprovalPolicy stores an m-of-n policy. allowed_voters is the list of
+    // voters allowed to vote and num_votes_required is the threshold required
+    // for approval
+    struct ApprovalPolicy has store, drop, copy {
+        allowed_voters: vector<address>,
+        num_votes_required: u64,
+    }
+
+    // This is a helper struct which allows us to store a map of proposal_type -> policy
+    // in the ApprovalPolicies struct
+    struct ApprovalPolicyMap has store, copy {
+        proposal_type: vector<u8>,
+        policy: ApprovalPolicy,
+    }
+
+    // Used to store onchain the current approval policy list. Sample mapping:
+    // [
+    //   "create_validator_owner" -> {allowed_voters: ["alice", "bob", "carol"], num_votes_required: 2},
+    //   "create_designated_dealer" -> {allowed_voters: ["bob", "carol", "dave"], num_votes_required: 1}
+    // ]
+    struct ApprovalPolicies has key {
+        approval_policies: vector<ApprovalPolicyMap>,
+    }
```

Comment:
Since we're storing the approval policies under a root address, the initialization (move_to) would require require the root signer for every new module trying to use the voting module. I wanted to avoid this since we want to avoid using the root signer as much as possible. If there is a way to initialize the struct under the root address without the root signer, then that would be great.